pymtl3/datatypes/bit_struct.py

```diff
@@ -0,0 +1,535 @@
+  )
+
+#-------------------------------------------------------------------------
+# _mk_repr_fn
+#-------------------------------------------------------------------------
+# Creates a __repr__ function based on fields. For example, if fields
+# contains two field x (Bits4) and y (Bits4), _mk_init_fn will return a
+# function that looks like the following:
+#
+# def __repr__( self ):
+#  return self.__class__.qualname__ + f'(x={self.x!r}, y={self.y!r})'
+
+def _mk_repr_fn( fields ):
+  return _create_fn(
+    '__repr__',
+    [ 'self' ],
+    [ 'return self.__class__.__qualname__ + f"(' +
+      ', '.join([ f'{f.name}={{self.{f.name}!r}}' for f in fields ]) +
+      ')"']
+  )
+
+#-------------------------------------------------------------------------
+# _mk_eq_fn
+#-------------------------------------------------------------------------
+# Creates a __eq__ function based on fields. By default it just compares
+# each field. For example, if fields contains two field x (Bits4) and y
+# (Bits4), _mk_init_fn will return a function that looks like the
+# following:
+#
+# def __eq__( self, other ):
+#   if other.__class__ is self.__class__:
+#     return (self.x,self.y,) == (other.x,other.y,)
+#   else:
+#     raise NotImplemented
+
+def _mk_eq_fn( fields ):
+  self_tuple  = _mk_tuple_str( 'self', fields )
+  other_tuple = _mk_tuple_str( 'other', fields )
+  return _create_fn(
+    '__eq__',
+    [ 'self', 'other' ],
+    [ f'return (other.__class__ is self.__class__) and {self_tuple} == {other_tuple}' ]
+  )
+
+#-------------------------------------------------------------------------
+# _mk_hash_fn
+#-------------------------------------------------------------------------
+# Creates a __hash__ function based on fields. By default it just hashes
+# all fields. For example, if fields contains two field x (Bits4) and y
+# (Bits4), _mk_init_fn will return a function that looks like the
+# following:
+#
+# def __hash__( self ):
+#   return hash((self.x,self.y,))
+
+def _mk_hash_fn( fields ):
+  self_tuple = _mk_tuple_str( 'self', fields )
+  return _create_fn(
+    '__hash__',
+    [ 'self' ],
+    [ f'return hash({self_tuple})' ]
+  )
+
+#-------------------------------------------------------------------------
+# _mk_mk_msg
+#-------------------------------------------------------------------------
+# __init__ doesn't perform casting and is usually used for constructing an
+# empty message. mk_msg does casting for Bits fields and preserve
+# list/BitStruct field, and is a class method
+#
+# @classmethod
+# def mk_msg(cls, x, y, some_struct_inst):
+#   return cls(Bits4(x), Bits8(y), some_struct_inst)
+
+def _mk_mk_msg( fields ):
+
+  # Register necessary types in _globals
+  _globals = {}
+
+  assign_strs = []
+  for f in fields:
+    type_ = f.type_
+    if isinstance( type_, list ) or is_bit_struct( type_ ):
+      assign_strs.append( f'{f.name}' )
+    else:
+      assert issubclass( type_, Bits )
+      _globals[ type_.__name__ ] = type_
```

Comment:
i don't think using type.__name__ is a good idea. If we have two types that has the same name but from different modules this wouldn't work. For example I want to use a `a.Point` and `b.Point` in my nested struct this wouldn't work. That why I rename all types to `_type_fname` in the dictionary but you seemed to have changed it

Reply:
i will add it back

Reply:
well if you had a test to catch this I won't remove it